### PR TITLE
AA-221: Prevent Paymaster 'postOp' function from reverting the entire bundle

### PR DESCRIPTION
**Pull request by forshtat**

### CODE REVIEW

File: test/entrypoint.test.ts [hunk 1]
```diff
@@ -1075,6 +1075,26 @@ describe('EntryPoint', function () {
...
+          initCode: getAccountInitCode(account3Owner.address, simpleAccountFactory),
+
+          verificationGasLimit: 3e6,
+          callGasLimit: 1e6
+        }, account3Owner, entryPoint)
+        const beneficiaryAddress = createAddress()
+        await entryPoint.handleOps([op], beneficiaryAddress)
+      })
+
+      it('should not revert when paymaster reverts with known EntryPoint error in postOp', async function () {
```
Comment (drortirosh) [inline test/entrypoint.test.ts:1111]: at minimum, should verify that the action was reverted (success=false, read the pm "reverted" data)

File: contracts/core/EntryPoint.sol [hunk 2]
```diff
@@ -662,9 +662,12 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuard,
                 if (context.length > 0) {
                     actualGasCost = actualGas * gasPrice;
                     if (mode != IPaymaster.PostOpMode.postOpReverted) {
-                        IPaymaster(paymaster).postOp{
+                        try IPaymaster(paymaster).postOp{
                             gas: mUserOp.verificationGasLimit
-                        }(mode, context, actualGasCost);
+                        }(mode, context, actualGasCost)
+                        {} catch (bytes memory reason) {
+                            revert(string(abi.encodePacked("AA96 postOp reverted", reason)));
```
Comment (drortirosh) [inline contracts/core/EntryPoint.sol]: this is problematic: you append binary encoding of custom error to a string.
(try to decode the error from the test..)
at a minimum: change string suffix to "...reverted: ", as we do with other reverts

File: test/entrypoint.test.ts [hunk 3]
```diff
@@ -1090,6 +1091,27 @@ describe('EntryPoint', function () {
...
+          callData: accountExecFromEntryPoint.data,
+          initCode: getAccountInitCode(account3Owner.address, simpleAccountFactory),
+
+          verificationGasLimit: 3e6,
+          callGasLimit: 1e6
+        }, account3Owner, entryPoint)
+        const beneficiaryAddress = createAddress()
+        const rcpt1 = await entryPoint.handleOps([op], beneficiaryAddress).then(async t => await t.wait())
+        const logs1 = await entryPoint.queryFilter(entryPoint.filters.UserOperationEvent(), rcpt1.blockHash)
+        expect(logs1[0].args.success).to.be.false
```
Comment (drortirosh) [inline test/entrypoint.test.ts:1129]: also the "postOp reverted" message

File: test/entrypoint.test.ts [hunk 4]
```diff
@@ -1090,6 +1091,27 @@ describe('EntryPoint', function () {
         await entryPoint.handleOps([op], beneficiaryAddress)
```
Comment (drortirosh) [inline test/entrypoint.test.ts]: handleOps doesn't revert, but you should check event side effects here too

File: test/entrypoint.test.ts (hunk 3 above)
Comment (forshtat) [reply test/entrypoint.test.ts:1129]: > also the "postOp reverted" message

There seems to be no place where `"postOp reverted"` error could be observed. We do not currently emit it as a part of any event, do we?

File: test/entrypoint.test.ts (hunk 4 above)
Comment (forshtat) [reply test/entrypoint.test.ts]: Added a check for `success == false` here. What other side effects do you want to check here?

File: test/entrypoint.test.ts (hunk 1 above)
Comment (forshtat) [reply test/entrypoint.test.ts:1111]: This is done.

File: contracts/core/EntryPoint.sol (hunk 2 above)
Comment (forshtat) [reply contracts/core/EntryPoint.sol]: Changed to revert with custom error that we can parse off-chain if we add the "postOp reverted" event (see next PR https://github.com/eth-infinitism/account-abstraction/pull/362)

File: test/testutils.ts [hunk 5]
```diff
@@ -172,6 +172,13 @@ export function decodeRevertReason (data: string, nullIfNoMatch = true): string
   } else if (methodSig === '0x4e487b71') {
     const [code] = ethers.utils.defaultAbiCoder.decode(['uint256'], dataParams)
     return `Panic(${panicCodes[code] ?? code} + ')`
+  } else if (methodSig === '0x8d6ea8be') {
+    const [reason] = ethers.utils.defaultAbiCoder.decode(['string'], dataParams)
+    return `CustomError("${reason as string}")`
```
Comment (drortirosh) [inline test/testutils.ts:177]: This `testutils` package was not meant to be a place for test-specific methods, but for "system-wide" reverts (yes, due to a lack of such a decoder in ethers itself)

**Review by drortirosh — APPROVED** [review]

**Pull request merged**